### PR TITLE
fix: skip unloaded iframes for all apis

### PR DESCRIPTION
**Pull request by straker**
Implements the webdriverjs fix into all apis. Playwright was the only api that could handle unloaded iframes without issues, so it didn't need to do anything special and thus can give different results than the other apis if there are unloaded iframes on the page.

This fix doesn't fix the issue for legacy mode. If we want to support legacy mode and unloaded iframes we should do it in a separate pr.

Closes https://github.com/dequelabs/axe-core-npm/issues/727

### CODE REVIEW

File: packages/webdriverio/test/axe-webdriverio.spec.ts [hunk 1]
```diff
@@ -850,6 +850,40 @@ describe('@axe-core/webdriverio', () => {
...
+          it('resets pageLoad timeout to user setting', async () => {
+            await client.url(`${addr}/lazy-loaded-iframe.html`);
+            client.setTimeout({ pageLoad: 500 })
+            const title = await client.getTitle();
+
+            const results = await new AxeBuilder({client})
+              .options({ runOnly: ['label', 'frame-tested'] })
+              .analyze();
+
+            const timeout = await client.getTimeouts();
```
Comment (WilcoFiers) [inline packages/webdriverio/test/axe-webdriverio.spec.ts:884]: Do you need to reset this in an after call?

**Review by WilcoFiers — DISMISSED** [review]
Couple small points

File: packages/puppeteer/src/utils.ts [hunk 2]
```diff
@@ -76,7 +76,16 @@ export async function assertFrameReady(frame: Frame): Promise<void> {
   // Check if the page is loaded.
   let pageReady = false;
   try {
-    pageReady = await frame.evaluate(pageIsLoaded);
+    // Puppeteer freezes on unloaded iframes. Set a race timeout in order to handle that.
+    // @see https://github.com/dequelabs/axe-core-npm/issues/727
+    const timeoutPromise = new Promise(resolve => {
+      setTimeout(() => {
+        resolve('timeout');
```
Comment (stephenmathieson) [inline packages/puppeteer/src/utils.ts:83]: Resolving with the value "timeout" is not at all idiomatic. If we time out, that is a sign of a problem, which is what the `reject` callback is meant to be used for. If we reject this promise (as discussed), we'd enter the `catch` block below (which _should_ throw the `Page/Frame is not ready` error).

**Review by stephenmathieson — DISMISSED** [review]
Let's reject promises rather than fulfilling them with an error.

File: packages/puppeteer/src/utils.ts (hunk 2 above)
Comment (dylanb) [reply packages/puppeteer/src/utils.ts:83]: we do not want to throw in these scenarios

File: packages/webdriverio/src/utils.ts [hunk 3]
```diff
@@ -98,6 +101,39 @@ export const axeSourceInject = async (
...
+      client.execute() on an unloaded iframe would cause
+      the code to hang indefinitely since it is using
+      Puppeteer which freezes on unloaded iframes. Set a
+      race timeout in order to handle that. Code taken
+      from our @axe-core/puppeteer utils function.
+      @see https://github.com/dequelabs/axe-core-npm/issues/727
+    */
+    const timeoutPromise = new Promise(resolve => {
+      setTimeout(() => {
+        resolve('timeout');
```
[the remark this answers is not in the document]
Comment (dylanb) [reply packages/webdriverio/src/utils.ts:121]: we really don't want to fail in these scenarios, we want to analyze the page content that we can analyze

File: packages/puppeteer/src/utils.ts (hunk 2 above)
Comment (stephenmathieson) [reply packages/puppeteer/src/utils.ts:83]: We do throw in these scenarios. See below.

File: packages/webdriverio/src/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephenmathieson) [reply packages/webdriverio/src/utils.ts:121]: We use the error as flow control. See below.

File: packages/webdriverio/test/axe-webdriverio.spec.ts (hunk 1 above)
Comment (straker) [reply packages/webdriverio/test/axe-webdriverio.spec.ts:884]: Good point, this is bleeding through the tests which could cause problems.

File: packages/webdriverio/src/utils.ts [hunk 4]
```diff
@@ -98,6 +101,33 @@ export const axeSourceInject = async (
...
+      setTimeout(() => {
+        reject();
+      }, FRAME_LOAD_TIMEOUT)
+    });
+    const executePromise = client.execute(() => {
+      return document.readyState === 'complete'
+    });
+    await Promise.race([timeoutPromise, executePromise]);
+  } catch {
+    throw new Error('Page/Frame is not ready');
```
Comment (dylanb) [inline packages/webdriverio/src/utils.ts:127]: Isn't part of the point of making these changes that we do not throw an exception?

**Review by dylanb — CHANGES_REQUESTED** [review]
What is the effect on an analyze call of throwing that exception? If it is going to cause the analyze call to throw, then I don't think that is the correct behavior

**Comment by stephenmathieson** [comment]
LGTM but I'll wait to approve until Dylan's question is answered.

File: packages/webdriverio/src/utils.ts (hunk 4 above)
Comment (straker) [reply packages/webdriverio/src/utils.ts:128]: We decided that the quickest way to resolve the bug was to return these iframes as `frame-tested` issues. How we want to handle unloaded / lazy loaded iframes is still a topic for discussion and being decided, but that shouldn't slow us down getting a fix out to the client.

**Review by stephenmathieson — APPROVED** [review]

**Pull request merged**